### PR TITLE
Quickstart SAP: Initial Draft

### PR DESCRIPTION
**Pull request by Loquacity**
https://github.com/SUSE/spacewalk/issues/10772

### CODE REVIEW

File: modules/quickstart-sap/pages/prepare-clients.adoc [hunk 1]
```diff
@@ -0,0 +1,42 @@
+[[quickstart-sap-setup]]
+= Setting Up
+
+Your SAP cluster requires several client systems.
+Prepare your clients on physical or virtual hardware, and ensure you have {sles}{nbsp}15 for SAP installation media ready on your {productname} Server.
+You will not be able to create an SAP cluster without the {sles} SAP extension, as it provides tooling specific to SAP.
+
+One of the key features of SAP is high availability of the cluster.
+Every component within an SAP cluster has redundancy and failover protection.
+When you are preparing your clients, ensure you have enough hardware and infrastructure to allow for this.
```
Comment (arbulu89) [inline modules/quickstart-sap/pages/prepare-clients.adoc:10]: I think we document this minimal requirement somewhere, but I don't honestly know

File: modules/quickstart-sap/pages/prepare-clients.adoc [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+
+
+.Procedure: Registering Clients with the {webui}
```
Comment (arbulu89) [inline modules/quickstart-sap/pages/prepare-clients.adoc:14]: Is this process used to register to SCC?

**Comment by Loquacity** [comment]
> Hi @Loquacity ,
> This initial version looks nice! Anyway, it barely talks about the formulas hehe So I guess the part where the formulas usage (and the pre-steps is missing, I have added them in the previous GH ticket).

Thanks! This will all end up in the configure.adoc section. I'll get started on that this week.

> 
> Let me know which is the step we have to take next.

I'll keep building on this PR for a little while, until we have all the steps included and fleshed out. Hoping to have that by the end of next week at the latest. From there, we can widen the tech review circle to a few others and start drilling down on the details.

File: modules/quickstart-sap/pages/prepare-clients.adoc (hunk 2 above)
Comment (Loquacity) [reply modules/quickstart-sap/pages/prepare-clients.adoc:14]: This registers the client systems to the SUMA Server. I don't think I have registering to the SCC in here, though, so I'll add that in.

File: modules/quickstart-sap/pages/prepare-clients.adoc (hunk 1 above)
Comment (Loquacity) [reply modules/quickstart-sap/pages/prepare-clients.adoc:10]: Found it in the SAP Guide. Added.

**Comment by Loquacity** [comment]
Hi @arbulu89 Ready for another review! There are some FIXME areas that I would like your feedback on, especially in modules/quickstart-sap/pages/configure.adoc.

File: modules/quickstart-sap/pages/prepare-clients.adoc [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+----
+- 'use_superseded:'
+      - '- module.run'
+----
+
+
+== Install Additional Disks for HANA
+
+
+For the clients that are going to run the HANA database, you will require an additional storage device.
```
Comment (arbulu89) [inline modules/quickstart-sap/pages/prepare-clients.adoc:73]: This is the most basic installation. Usually, we need multiple disks (up to 10).
Here some more info:
https://documentation.suse.com/sbp/all/html/SBP-SAP-AzureSolutionTemplates/index.html

File: modules/quickstart-sap/pages/configure.adoc [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+----
+systemctl restart salt-master.service
+----
+
+
+
+== Enable and Configure the HANA Formula
+
+In the {productname} {webui}, navigate to menu:Systems[System List] and click the ``FIXME`` client.
+This is the primary client that you want to install the HANA database on.
```
[the remark this answers is not in the document]
Comment (Loquacity) [reply modules/quickstart-sap/pages/configure.adoc:39]: So I guess that means you install the formula on both clients? Or do you just pick one as 'primary' and input details for both?

File: modules/quickstart-sap/pages/prepare-clients.adoc [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+== Download the SAP Installation Software
+
+Download the {sles}{nbsp}15 with SAP installation media from {scc}, and save a copy on each clients.
+Ensure you have saved it in the same file path on each client.
+
+
+
+== Configure Clients to Use Latest ``module.run``
+
+// Not sure what this is. --LKB 2020-05-20
```
[the remark this answers is not in the document]
Comment (Loquacity) [reply modules/quickstart-sap/pages/prepare-clients.adoc:59]: Which config file is being edited?

File: modules/quickstart-sap/pages/configure.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (arbulu89) [reply modules/quickstart-sap/pages/configure.adoc]: No, the formula is installed in the SUMA master. What I mean is that we can run the formula execution once for the 2 clients. 
We configure the pillar file to say to the execution to install HANA primary in one node and the secondary in the 2nd node, but everything is done during the same salt execution

File: modules/quickstart-sap/pages/prepare-clients.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (arbulu89) [reply modules/quickstart-sap/pages/prepare-clients.adoc]: `/etc/salt/master` (`/etc/salt/minion` if you run salt as local, what doesn't apply with SUMA I guess)

File: modules/quickstart-sap/pages/configure.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (Loquacity) [reply modules/quickstart-sap/pages/configure.adoc]: So I've found the packages in the SUSE gh repo, is there a package that people can download to install them? I thought they were in the ordinary repos, but they don't seem to be:

image

[the remark this answers is not in the document]
Comment (Loquacity) [reply modules/quickstart-sap/pages/configure.adoc]: Also: the repo given in the HANA readme is out of date: 

image

[the remark this answers is not in the document]
Comment (arbulu89) [reply modules/quickstart-sap/pages/configure.adoc]: Hi @Loquacity ,
What repository are you using? These 2 packages should be available in SLES4SAP subsciption. The truth is that using SUMA doesn't make much sense maybe how we install the formulas. The thing is that the formulas are used to change SLES4SAP machines, but they might be executed from a non SLES4SAP (SUMA master) for example, so maybe this machine doesn't have SLES4SAP subscription (most probably).

@diegoakechi What do you think here?

And yes, the HANA formula project readme is outdated..

[the remark this answers is not in the document]
Comment (diegoakechi) [reply modules/quickstart-sap/pages/configure.adoc]: > @diegoakechi What do you think here?
@Loquacity,

What was aligned between Joe and Christian (PM of the related products) was that the salt formulas would be shipped on the SLES for SAP repositories and SUMA has a repository mirroring mechanism where you can install a package from the SLES for SAP repo (if you have a SLES for SAP subscription) into the SUMA server. 
I am not familiar with how is it done, but you need for sure to have a SLES for SAP subscription/repositories configured on SUMA for that.

It was documented here: https://confluence.suse.com/display/haslesforsap/%5BREVIEW%5D+Terraform+Deployment+Project+Support#id-[REVIEW]TerraformDeploymentProjectSupport-SALTformulasDelivery

[the remark this answers is not in the document]
Comment (Loquacity) [reply modules/quickstart-sap/pages/configure.adoc]: Ok, thanks for this info. I might need to do some further digging to work this out. I'll let you know how I go!

[the remark this answers is not in the document]
Comment (Loquacity) [reply modules/quickstart-sap/pages/configure.adoc]: Looks like the packages are in `SUSE Package Hub 15 SP1 x86_64`.

[the remark this answers is not in the document]
Comment (renner) [reply modules/quickstart-sap/pages/configure.adoc]: > SUMA has a repository mirroring mechanism where you can install a package from the SLES for SAP repo (if you have a SLES for SAP subscription) into the SUMA server.

This is an interesting idea, but I fear that it is not going to work yet: while SUSE Manager is mirroring the repos, AFAIK it is not (yet) supported to manage the SUSE Manager Server itself to assign repos and to install packages. Is that maybe how they wanted this to work?

[the remark this answers is not in the document]
Comment (diegoakechi) [reply modules/quickstart-sap/pages/configure.adoc]: > > SUMA has a repository mirroring mechanism where you can install a package from the SLES for SAP repo (if you have a SLES for SAP subscription) into the SUMA server.
> 
> This is an interesting idea, but I fear that it is not going to work yet: while SUSE Manager is mirroring the repos, AFAIK it is not (yet) supported to manage the SUSE Manager Server itself to assign repos and to install packages. Is that maybe how they wanted this to work?

@renner better ask Joe as he has aligned this with Christian on the PM side. I never checked this myself, but I understood he was sure it worked this way.

[the remark this answers is not in the document]
Comment (Loquacity) [reply modules/quickstart-sap/pages/configure.adoc]: image

File: modules/quickstart-sap/pages/configure.adoc [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+[IMPORTANT]
+====
+The order that you enable and configure the formulas is important.
+You must enable, configure, and apply the HANA formula first.
+Then you can enable, configure, and apply the cluster formula.
+If you perform these steps in the wrong order, your SAP installation will fail.
+====
+
+
+To install the SAP formulas, you need the ``SUSE Package Hub 15 SP1 x86_64`` product from {scc}.
```
Comment (diegoakechi) [inline modules/quickstart-sap/pages/configure.adoc:22]: @Loquacity We released it on package Hub as a beta version. The official packages will be part of the SLES for SAP official repositories, so enabling packagehub will not be needed in a near future. On SLES for SAP 15 SP2 the packages are already available as we have ECOs as WIP to backport this to older SLES for SAP 12 SP3 and onwards.

File: modules/quickstart-sap/pages/prepare-clients.adoc [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+Your SAP cluster requires several client systems.
+Prepare your clients on physical or virtual hardware, and ensure you have {sles}{nbsp}15 for SAP installation media ready on your {productname} Server.
+You will not be able to create an SAP cluster without the {sles} SAP extension, as it provides tooling specific to SAP.
+
+One of the key features of SAP is high availability of the cluster.
+Every component within an SAP cluster has redundancy and failover protection.
+When you are preparing your clients, ensure you have enough hardware and infrastructure to allow for this.
+For more information about hardware requirements, see https://documentation.suse.com/sles-sap/15-SP1/html/SLES4SAP-quick/cha-s4s-plan.html#sec-s4s-hardware[].
+
+For more information about the clients you need to set up for an SAP cluster, see https://documentation.suse.com/sbp/all/html/SBP-SAP-AzureSolutionTemplates/index.html[].
```
Comment (diegoakechi) [inline modules/quickstart-sap/pages/prepare-clients.adoc:13]: @Loquacity This link is specific for Azure Templates, but the SAP and Cluster requirements varies depending on the platform. I would point it to the SUSE Best Practices page, where we have many SAP related documents: https://documentation.suse.com/sbp/all/

File: modules/quickstart-sap/pages/prepare-clients.adoc [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+----
+
+
+== Install Additional Disks for HANA
+
+
+For the clients that are going to run the HANA database, you will require an additional storage device.
+This device is used to store files required by HANA, which are located in the [path]``/hana/`` directory.
+
+We recommend that this storage device be at least 20{nbsp}GB.
```
Comment (diegoakechi) [inline modules/quickstart-sap/pages/prepare-clients.adoc:72]: I would recommend to point the user to check the official SAP documentation about that, as HANA has specific requirements and SAP can change it at any time.

File: modules/quickstart-sap/pages/prepare-server.adoc [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+Before you start you need to install the {productname} Server.
+The method for installing the {productname} Server will vary depending on your hardware and environment.
+
+For more information about installing the {productname} Server, see xref:installation:install-intro.adoc[].
+
+
+
+You need access to {sles}{nbsp}15 for SAP installation media.
+You will not be able to create an SAP cluster without the {sles} SAP extension, as it provides tooling specific to SAP.
```
Comment (diegoakechi) [inline modules/quickstart-sap/pages/prepare-server.adoc:12]: @Loquacity SLES for SAP is technically a SUSE Base Product, not an extension.

**Comment by Loquacity** [comment]
After some trying, I still have not managed to actually install an SAP formula. I've cleaned up the FIXMEs for now, which will allow us to publish this as a draft doc. That way we can get some more eyes on it.

**Comment by Loquacity** [comment]
@jcayouette can you please add this as a book for me?

**Comment by diegoakechi** [comment]
@Loquacity The formulas are available only on SLES for SAP 15 SP2 at the moment, as we are waiting the approval of the https://jira.suse.com/browse/ECO-1965.

Also, accordingly to @renner and @paususe there is no way to use the package from the SLES for SAP into the SUMA server machine, so here we need to decide how to make it appear there. At the moment, you would need to copy the files manually from the SLES for SAP machine into the SUMA server, what I think is a very bad idea. @renner and @paususe would discuss with PM about or shipping these packages on SUMA server too or maybe some other option. @renner @paususe any news about that?

**Comment by Loquacity** [comment]
> @Loquacity The formulas are available only on SLES for SAP 15 SP2 at the moment, as we are waiting the approval of the https://jira.suse.com/browse/ECO-1965.
> 
> Also, accordingly to @renner and @paususe there is no way to use the package from the SLES for SAP into the SUMA server machine, so here we need to decide how to make it appear there. At the moment, you would need to copy the files manually from the SLES for SAP machine into the SUMA server, what I think is a very bad idea. @renner and @paususe would discuss with PM about or shipping these packages on SUMA server too or maybe some other option. @renner @paususe any news about that?

Hrm, that changes things. I'm not sure there's much point shipping this doc if you can't easily install the formulas. I think we need @paususe to make a call here.

**Comment by paususe** [comment]
> Also, accordingly to @renner and @paususe there is no way to use the package from the SLES for SAP into the SUMA server machine, so here we need to decide how to make it appear there. At the moment, you would need to copy the files manually from the SLES for SAP machine into the SUMA server, what I think is a very bad idea. @renner and @paususe would discuss with PM about or shipping these packages on SUMA server too or maybe some other option. @renner @paususe any news about that?

I'm working with PM about this, we'll find a solution but I don't have a timeline yet

**Comment by Loquacity** [comment]
> > Also, accordingly to @renner and @paususe there is no way to use the package from the SLES for SAP into the SUMA server machine, so here we need to decide how to make it appear there. At the moment, you would need to copy the files manually from the SLES for SAP machine into the SUMA server, what I think is a very bad idea. @renner and @paususe would discuss with PM about or shipping these packages on SUMA server too or maybe some other option. @renner @paususe any news about that?
> 
> I'm working with PM about this, we'll find a solution but I don't have a timeline yet

OK, let's put this on ice for the moment then, and we can pick it back up when we have a good solution. Thanks everyone.

File: modules/quickstart-sap/pages/qs-sap-intro.adoc [hunk 10]
```diff
@@ -3,7 +3,8 @@
 
 **Publication Date:** {docdate}
 
-This guide shows you the fastest way to get {productname} up and running to manage an SAP cluster.
-It guides you through setting up a single {productname} Server, installing the SAP cluster, registering clients, and setting up a content lifecycle for managing them.
+This guide shows you how to use {productname} to install and configure an SAP cluster.
```
Comment (arbulu89) [inline modules/quickstart-sap/pages/qs-sap-intro.adoc:6]: I should be `configure a SAP cluster`, right?
And I would add actually add `...configure a SAP cluster with High Availability feature` or similar

File: modules/quickstart-sap/pages/prepare-clients.adoc [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+One of the key features of SAP is high availability of the cluster.
+Every component within an SAP cluster has redundancy and failover protection.
+When you are preparing your clients, ensure you have enough hardware and infrastructure to allow for this.
+For more information about hardware requirements, see https://documentation.suse.com/sles-sap/15-SP1/html/SLES4SAP-quick/cha-s4s-plan.html#sec-s4s-hardware[].
+
+For more information about the clients you need to set up for an SAP cluster, see https://documentation.suse.com/sbp/all[].
+
+
+
+== Register Clients to the {SCC}
```
Comment (arbulu89) [inline modules/quickstart-sap/pages/prepare-clients.adoc:17]: Maybe it's worth mentioning that they need to register to a SLES4SAP subscription, but this information might be redundant, as you already mentioned that the machines must of this version

File: modules/quickstart-sap/pages/prepare-clients.adoc [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+SUSEConnect -r <Registration_Number> -e <registered_email_address>
+----
+
+You can get your registration number by logging in to {scc} from your browser.
+
+For more information about registering {productname} with {scc}, see xref:installation:general-requirements.adoc[].
+
+
+
+== Create the Cluster
```
Comment (arbulu89) [inline modules/quickstart-sap/pages/prepare-clients.adoc:31]: Agree with @MalloZup , I would rename this to `Configure the cluster machines`

File: modules/quickstart-sap/pages/prepare-clients.adoc [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+The shared disk can be physical hardware connected by ethernet, or you can set up a virtual disk and access it with iSCSI.
+
+If you use a virtual disk, consider hosting it on a separate system.
+Do not use a client machine to host the shared storage disk.
+
+
+
+== Download the SAP Installation Software
+
+Download the {sles}{nbsp}15 with SAP installation media from {scc}, and save a copy on each clients.
```
Comment (arbulu89) [inline modules/quickstart-sap/pages/prepare-clients.adoc:50]: This paragraph is incorrect. Here we have to state that the users must download and have the **SAP** installation software in the client machines. The files can be stored in a NFS share too.

Here we need to have specific SAP software, that is obtained from the SAP download center (not SUSE).
- For HANA, we need SAP HANA platform (or any similar product).
- For NETWAEVER, S4/HANA things get a more bit difficult, as this requires more software than only one file.

I don't know how we could describe this, but it's not easy.
I don't know how the other SLES4SAP guides mention this part honestly. 
But we have to make sure that this is SAP software, and it's not provided by SUSE

File: modules/quickstart-sap/pages/prepare-clients.adoc [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+Ensure you have saved it in the same file path on each client.
+
+
+
+== Configure Clients to Use Latest ``module.run``
+
+Each client needs to be configured to use the latest version of ``module.run``.
+On the {productname} Server, open the ``/etc/salt/master`` configuration file and add or amend this line:
```
Comment (arbulu89) [inline modules/quickstart-sap/pages/prepare-clients.adoc:58]: If I'm not wrong, this `module.run` configuration must be done in each of the clients. It cannot be done in the `master` machine. Any of the `salt` developers could confirm this?
If this is the case, the file to change would be `/etc/salt/minion`

PD: Now, if I'm not wrong we ship salt 3000, and this version might use already correctly this option

**Review by arbulu89 — COMMENTED** [review]
I have added some important comments to the guide.

Finally, as @MalloZup said, we need to think if we have to add examples of the pillar files configuration.
I have found out in the past, that if the users need to do this alone, without any guidance, they will fail most probably.

Anyway, it looks like @tomschr  is working on something for this.

**Comment by tomschr** [comment]
> Anyway, it looks like @tomschr is working on something for this.

Yes, if you are referring on the how to use Salt formulas. :wink:  The first draft (including PDF) can be found in this PR: SUSE/doc-slesforsap#41

**Comment by Loquacity** [comment]
@arbulu89 you are looking at a very old version, and many of these things have been resolved. I'll open a new PR now, that might make it a bit easier.

**Comment by Loquacity** [comment]
Please check https://github.com/uyuni-project/uyuni-docs/pull/559

**Pull request merged**